### PR TITLE
Fix locations for defs

### PR DESCRIPTION
**Pull request by mlutze**
Fixes https://github.com/flix/flix/issues/2469

For consistency, these same steps should be followed for all definitions.

### CODE REVIEW

**Comment by magnus-madsen** [comment]
Do you want to do the others in this PR too?

**Comment by magnus-madsen** [comment]
I would have given it to a student programmer, but I think its too complicated.

**Comment by mlutze** [comment]
> Do you want to do the others in this PR too?

Sure.

**Comment by mlutze** [comment]
It appears that Enums, Opaque Types, and Type Aliases all carry the locations associated with the full declarations, so nothing to do there.

**Comment by magnus-madsen** [comment]
I think we need to more carefully look through the changes.

File: main/src/ca/uwaterloo/flix/api/lsp/LocationLink.scala [hunk 1]
```diff
@@ -33,7 +33,7 @@ object LocationLink {
     val originSelectionRange = Range.from(loc)
     val targetUri = sym.loc.source.name
     val targetRange = Range.from(sym.loc)
-    val targetSelectionRange = Range.from(defDecl.spec.loc)
+    val targetSelectionRange = Range.from(defDecl.sym.loc)
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/api/lsp/LocationLink.scala:36]: For example here, should we not use the entire loc?

File: main/src/ca/uwaterloo/flix/api/lsp/LocationLink.scala [hunk 2]
```diff
@@ -45,7 +45,7 @@ object LocationLink {
     val originSelectionRange = Range.from(loc)
     val targetUri = sym.loc.source.name
     val targetRange = Range.from(sym.loc)
-    val targetSelectionRange = Range.from(sigDecl.spec.loc)
+    val targetSelectionRange = Range.from(sigDecl.sym.loc)
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/api/lsp/LocationLink.scala:48]: also here?

File: main/src/ca/uwaterloo/flix/api/lsp/LocationLink.scala [hunk 3]
```diff
@@ -90,9 +90,9 @@ object LocationLink {
     */
   def fromInstance(instance: TypedAst.Instance, originLoc: SourceLocation): LocationLink = {
     val originSelectionRange = Range.from(originLoc)
-    val targetUri = instance.loc.source.name
-    val targetRange = Range.from(instance.loc)
-    val targetSelectionRange = Range.from(instance.loc)
+    val targetUri = instance.sym.loc.source.name
+    val targetRange = Range.from(instance.sym.loc)
+    val targetSelectionRange = Range.from(instance.sym.loc)
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/api/lsp/LocationLink.scala:95]: and here

File: main/src/ca/uwaterloo/flix/language/ast/Symbol.scala [hunk 4]
```diff
@@ -178,6 +178,13 @@ object Symbol {
     new TypeAliasSym(ns.parts, ident.name, ident.loc)
   }
 
+  /**
+    * A union type for [[DefnSym]]s and [[SigSym]]s
+    */
+  sealed trait DefnSymOrSigSym {
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/ast/Symbol.scala:184]: Why is this necessary? I don't like this proliferation of multiple inheritance.

**Comment by magnus-madsen** [comment]
I think you should build a JAR and fire up VSCode and then manually test each feature impacted by the changes. Then, once you feel satisfied, I can take a more detailed look. Sounds like a plan?

File: main/src/ca/uwaterloo/flix/api/lsp/LocationLink.scala (hunk 1 above)
Comment (mlutze) [reply main/src/ca/uwaterloo/flix/api/lsp/LocationLink.scala:36]: Yes, my goal was avoiding regression, so in these places where I don't really know what's going on, I maintained the same behavior. I'll take another look at these.

File: main/src/ca/uwaterloo/flix/language/ast/Symbol.scala (hunk 4 above)
Comment (mlutze) [reply main/src/ca/uwaterloo/flix/language/ast/Symbol.scala:184]: We have a place that returns either a DefnSym or a SigSym, based on a lookup. I'll see if it can be turned into an enum. Otherwise, more generally would it be useful to have a `Located` trait, so it's more like a type class than a union of types?

**Comment by mlutze** [comment]
> I think you should build a JAR and fire up VSCode and then manually test each feature impacted by the changes. Then, once you feel satisfied, I can take a more detailed look. Sounds like a plan?

Sure. I was thinking it would be nice to maintain some sort of "test card" for the plugin, a file or set of files containing all the features we want to check.

File: main/src/ca/uwaterloo/flix/language/ast/Symbol.scala (hunk 4 above)
Comment (mlutze) [reply main/src/ca/uwaterloo/flix/language/ast/Symbol.scala]: I replaced the DefOrSig union trait with a disjoint union.

**Comment by mlutze** [comment]
I'm not able to find any features working on master but not working on the feature branch.
